### DISCUSSION TITLE
how to solve "error: failed to select a version for `libsqlite3-sys`"

### DISCUSSION THREAD

**Discussion by mobergmann** [maxcountryman/axum-login#95] [Q&A] [answered]
When I run `cargo build` I get the error
```
    Updating crates.io index
error: failed to select a version for `libsqlite3-sys`.
    ... required by package `sqlx-core v0.6.0`
    ... which satisfies dependency `sqlx-core = "^0.6.0"` of package `sqlx v0.6.0`
    ... which satisfies dependency `sqlx = "^0.6"` of package `axum-login v0.6.0`
    ... which satisfies dependency `axum-login = "^0.6.0"` of package `backend v0.1.0 (/home/mobergmann/Source/sp-backend)`
versions that meet the requirements `^0.24.1` are: 0.24.2, 0.24.1

the package `libsqlite3-sys` links to the native library `sqlite3`, but it conflicts with a previous package which links to `sqlite3` as well:
package `libsqlite3-sys v0.26.0`
    ... which satisfies dependency `libsqlite3-sys = "^0.26.0"` of package `sqlx-sqlite v0.7.1`
    ... which satisfies dependency `sqlx-sqlite = "=0.7.1"` of package `sqlx v0.7.1`
    ... which satisfies dependency `sqlx = "^0.7.1"` of package `backend v0.1.0 (/home/mobergmann/Source/sp-backend)`
Only one package in the dependency graph may specify the same links value. This helps ensure that only one copy of a native library is linked in the final binary. Try to adjust your dependencies so that only one package uses the links ='libsqlite3-sys' value. For more information, see https://doc.rust-lang.org/cargo/reference/resolver.html#links.

failed to select a version for `libsqlite3-sys` which could resolve this conflict
```
I am still new to axum and rust, so maybe I am overseeing something.  
Does anyone know how to fix this issue? I already tried to manually specify the version of `libsqlite3-sys` to `0.24.2` and `0.24.1`, but without success. I also tried to downgrade a bunch of packages, also without success.

my dependencies from the Cargo.toml
```
[dependencies]
argon2 = "0.5.2"
axum = "0.6.20"
axum-login = { version = "0.6.0", features = ["sqlx", "sqlite"] }
rand = "0.8.5"
serde = "1.0.188"
serde_json = "1.0.106"
sqlx = { version = "0.7.1", features = ["sqlite"] }
tokio = { version = "1.32.0", features = ["full"] }
```

I also tried to run the  dependencies from the Cargo.tom [from the examples](https://github.com/maxcountryman/axum-login/blob/main/examples/sqlite/Cargo.toml), but I get the error ``error: failed to select a version for `libsqlite3-sys`:
```
[dependencies]
axum = "0.6.0"

[dependencies.axum-login]
path = "../../axum-login"
features = ["sqlite"]

[dependencies.rand]
version = "0.8.5"
features = ["min_const_gen"]

[dependencies.sqlx]
version = "0.7"
default-features = false
features = ["runtime-tokio-rustls", "sqlite"]

[dependencies.tokio]
version = "1.0"
features = ["full"]
```

**Comment by maxcountryman** [comment]
This has to do with `async-sqlx-session` not being updated to use `sqlx` 0.7.x.

For now, you can use my fork:

```toml
[dependencies.async-sqlx-session]
 git = "https://github.com/maxcountryman/async-sqlx-session"
 branch = "main"
```

**Reply by mobergmann** [reply]
I cant get it working. Can you maybe provide a working dependencies set, based on the one I posted above?
I tried adding it and replacing sqlx with it. Both didn't work.

**Reply by mobergmann** [reply]
@maxcountryman sorry for pinging, but I'm really lost right now

**Reply by maxcountryman** [reply, marked as answer]
Sorry you're having trouble here. I'm working on some separate, upstream changes and my goal is to make the third party dependencies much easier to wrangle on the user side.

For now, you'll have to use the version of sqlx that your axum-login requires (so that would be 0.6.x I think; make sure axum-session requires the same).

**Comment by maxcountryman** [comment]
This should no longer be an issue now that we've redesigned the crate. Please let us know if you run into any additional issues.
